dali/operators/audio/nonsilence_op.h

```diff
@@ -0,0 +1,202 @@
+}
+
+
+/**
+ * Performs nonsilent region detection for single sample
+ * @return (begin_idx, length)
+ */
+template<typename InputType>
+std::pair<int, int>
+DetectNonsilenceRegion(Tensor<CPUBackend> &intermediate_buffer, const Args<InputType> &args) {
+  RunKernel(args.input, intermediate_buffer, {args.window_length, args.reset_interval});
+  auto dbs = view_as_tensor<float>(intermediate_buffer);
+  kernels::signal::DecibelCalculator<float> dbc(10.f, args.reference_max ? max(dbs)
+                                                                         : args.reference_power);
+  return LeadTrailThresh(make_cspan(dbs.data, dbs.num_elements()), dbc.db2signal(args.cutoff_db));
+}
+
+}  // namespace detail
+
+template<typename Backend>
+class NonsilenceOperator : public Operator<Backend> {
+ public:
+  ~NonsilenceOperator() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperator);
+
+ protected:
+  explicit NonsilenceOperator(const OpSpec &spec) :
+          Operator<Backend>(spec) {}
+
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  void AcquireArgs(const OpSpec &spec, const workspace_t<Backend> &ws) {
+    this->GetPerSampleArgument(cutoff_db_, "cutoff_db", ws);
+    this->GetPerSampleArgument(reference_power_, "reference_power", ws);
+    window_length_ = spec.GetArgument<int>("window_length", &ws);
+    auto input_type = ws.template InputRef<Backend>(0).type().id();
+    // If input type is not floating point, there's no need for reset interval
+    reset_interval_ = IsFloatingPoint(input_type) ? spec.GetArgument<int>("reset_interval", &ws)
+                                                  : -1;
+  }
+
+
+  std::vector<float> cutoff_db_;
+  std::vector<float> reference_power_;
+  int window_length_ = -1;
+  int reset_interval_ = -1;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+
+class NonsilenceOperatorCpu : public NonsilenceOperator<CPUBackend> {
+ public:
+  explicit NonsilenceOperatorCpu(const OpSpec &spec) :
+          NonsilenceOperator<CPUBackend>(spec) {
+    intermediate_buffers_.resize(num_threads_);
+  }
+
+
+  ~NonsilenceOperatorCpu() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperatorCpu);
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override;
+
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  template<typename InputType>
+  void RunImplTyped(workspace_t<CPUBackend> &ws) {
+    const auto &input = ws.template InputRef<CPUBackend>(0);
+    auto &output_begin = ws.OutputRef<CPUBackend>(0);
+    auto &output_length = ws.OutputRef<CPUBackend>(1);
+    auto &tp = ws.GetThreadPool();
+
+    for (int sample_id = 0; sample_id < batch_size_; sample_id++) {
+      tp.DoWorkWithID(
+              [&, sample_id](int thread_id) {
+                  detail::Args<InputType> args;
+                  args.input = view<const InputType, 1>(input[sample_id]);
```

Comment:
I believe input should not be part of Args. And also, Args should not be typed (possible if you decouple it from the input)

Reply:
As we discussed it while ago, in DALI (especially in kernels) input and args are treated almost the same way. Is there particular reason not to do so here? I don't see any major problems that might occur by making input one of the args. After all, both input and args are the coefficient, based on which output is calculated

Reply:
I will reiterate that especially in kernels this is not true.